File: prototype/rails/src/app/models/page_view_count.rb
# One route pattern's traffic on one day, on one side of the marketplace.
# Rolled up at response time rather than read back from a row per hit: the
# table grows with routes and days, not with traffic.
class PageViewCount < ApplicationRecord
  prefixed_id :pvc

  SITES = %w[shop seller admin].freeze

  validates :site, inclusion: { in: SITES }
  validates :path_pattern, presence: true
  validates :day, presence: true

  # Rolls one hit into its (site, path_pattern, day) row in one statement,
  # with no read first: the unique index is what makes the first hit of a
  # day an insert and every later one an increment.
  def self.record!(path_pattern:, at: Time.current)
    upsert(
      {
        id: PrefixedUlid.generate(:pvc),
        site: PageView.site_for(path_pattern),
        path_pattern: path_pattern,
        day: PageView.day(at),
        count: 1
      },
      unique_by: %i[site path_pattern day],
      on_duplicate: Arel.sql("count = count + 1")
    )
  end

  # Every day that saw traffic, newest first.
  def self.by_day
    group(:day).sum(:count).sort_by { |day, _count| day }.reverse
  end

  # Every route pattern that saw traffic, busiest first.
  def self.by_pattern
    group(:site, :path_pattern).sum(:count).sort_by { |(site, pattern), count| [ -count, site, pattern ] }
  end

  # How many hits a run of days holds, folded in one statement.
  def self.total_for(days)
    where(day: days).sum(:count)
  end
end

File: prototype/php/src/app/Models/PageViewCount.php
<?php

declare(strict_types=1);

namespace App\Models;

use App\Domain\Analytics\PageViewWeek;
use App\Models\Concerns\HasPrefixedUlid;
use Database\Factories\PageViewCountFactory;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Attributes\Scope;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Override;

/**
 * One row per (site, path_pattern, day): a request's route pattern rolled up
 * at response time rather than logged per hit, so the table grows with
 * routes and days rather than with traffic. The table lives in the
 * analytics store (config/database.php), a SQLite file of its own next to
 * the commerce database.
 * {@see \App\Analytics\Analytics::flush()} is the only writer, through an
 * upsert on this table's unique triple.
 */
#[Fillable(['site', 'path_pattern', 'day', 'count'])]
class PageViewCount extends Model
{
    /** @use HasFactory<PageViewCountFactory> */
    use HasFactory;

    use HasPrefixedUlid;

    protected $connection = 'analytics';

    public static function idPrefix(): string
    {
        return 'pvc';
    }

    /**
     * @return array<string, string>
     */
    #[Override]
    protected function casts(): array
    {
        return [
            'count' => 'integer',
        ];
    }

    /** @param Builder<$this> $query */
    #[Scope]
    protected function inWeek(Builder $query, PageViewWeek $week): void
    {
        $query->whereBetween('day', [$week->firstDay, $week->lastDay]);
    }

    public static function totalForWeek(PageViewWeek $week): int
    {
        return (int) self::query()->inWeek($week)->sum('count');
    }

    /**
     * Every day inside the window that saw traffic, newest first.
     *
     * @return list<array{day: string, count: int}>
     */
    public static function totalsByDay(PageViewWeek $week): array
    {
        return array_values(self::query()
            ->inWeek($week)
            ->select('day')
            ->selectRaw('sum(count) as count')
            ->groupBy('day')
            ->orderByDesc('day')
            ->get()
            ->map(fn (self $row): array => ['day' => $row->day, 'count' => (int) $row->count])
            ->all());
    }

    /**
     * Every route pattern that has ever seen traffic, busiest first.
     *
     * @return list<array{site: string, pathPattern: string, count: int}>
     */
    public static function totalsByPattern(): array
    {
        return array_values(self::query()
            ->select('site', 'path_pattern')
            ->selectRaw('sum(count) as count')
            ->groupBy('site', 'path_pattern')
            ->orderByDesc('count')
            ->orderBy('site')
            ->orderBy('path_pattern')
            ->get()
            ->map(fn (self $row): array => [
                'site' => $row->site,
                'pathPattern' => $row->path_pattern,
                'count' => (int) $row->count,
            ])
            ->all());
    }
}
